pkg/client/heartbeater/heartbeater.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package heartbeater
+
+import (
+	"context"
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/codenotary/immudb/pkg/logger"
+	"github.com/golang/protobuf/ptypes/empty"
+	stdos "os"
+	"time"
+)
+
+const KeepAliveInterval = time.Second * 1
+
+type heartBeater struct {
+	sessionID     string
+	logger        logger.Logger
+	serviceClient schema.ImmuServiceClient
+	done          chan bool
+	t             *time.Ticker
+}
+
+func NewHeartBeater(sessionID string, sc schema.ImmuServiceClient) *heartBeater {
+	return &heartBeater{
+		sessionID:     sessionID,
+		logger:        logger.NewSimpleLogger("immuclient", stdos.Stdout),
+		serviceClient: sc,
+		done:          make(chan bool),
+		t:             time.NewTicker(KeepAliveInterval),
+	}
+}
+
+func (hb *heartBeater) KeepAlive() (err error) {
+	go func() error {
+		for {
+			select {
+			case <-hb.done:
+				return nil
+			case t := <-hb.t.C:
+				hb.logger.Debugf("keep alive for %d at %s\n", hb.sessionID, t.String())
+				err = hb.keepAliveRequest()
+				if err != nil {
+					return err
```

Comment:
This is pretty dangerous - we're exiting the internal goroutine here, noone gets informed about this error but the heartbeat timer stops working.

It would be better to just log the error but continue with the next tick.

Reply:
Good point, an alternative is to log but check for an heartbeater error on a client interceptor. What do you think?

Reply:
I think simply logging the error here would be enough. From the client perspective the only reason it sends this keepalives in a loop is to prove to the server that its still alive, not to diagnose the connection. I wouldn't interfere with other operations, and if there's a critical error with the connection, any other GRPC call should fail too.

I'm not sure if there's a simple scenario where Keepalive fails but other GRPC calls succeed. Only if the server fails to process keepalive internally (so maybe new client with an old server that do not have keepalive?). I think that it's still better to just ignore keepalive errors in such situation.